File: src/main/java/features/reports/pdf/DataTableReportToPdfElementConverter.java
package features.reports.pdf;

import com.itextpdf.io.font.constants.StandardFonts;
import com.itextpdf.kernel.font.PdfFont;
import com.itextpdf.kernel.font.PdfFontFactory;
import com.itextpdf.layout.element.Cell;
import com.itextpdf.layout.element.IBlockElement;
import com.itextpdf.layout.element.Paragraph;
import com.itextpdf.layout.element.Table;
import com.itextpdf.layout.property.UnitValue;
import core.algorithm.Algorithm;
import core.data.Data;
import core.data.report.AlgorithmReport;
import core.data.report.DataTableAlgorithmReport;
import features.reports.PdfElementData;

import java.io.IOException;
import java.util.Collection;
import java.util.Dictionary;

/**
 * {@link Algorithm} that decorates report creator and generates {@link PdfElementData} from decorated {@link Algorithm} output
 * @param <I> Inner {@link Algorithm} input
 */
public class DataTableReportToPdfElementConverter<I extends Data> implements Algorithm<I, PdfElementData> {

    private Algorithm<I, AlgorithmReport> reportCreator;

    public DataTableReportToPdfElementConverter(Algorithm<I, AlgorithmReport> reportCreator) {
        this.reportCreator = reportCreator;
    }

    @Override
    public PdfElementData execute(I input) {
        var report = reportCreator.execute(input);
        return new PdfElementData(execute(report));
    }

    private IBlockElement execute(AlgorithmReport report) {
        IBlockElement generatedReport;

        switch (report.getMirfReportType()) {
            case DataTable:
                generatedReport = generatePdfForTableReport((DataTableAlgorithmReport) report);
                break;
            default:
                throw new RuntimeException(String.format("Cannot generate report for %s", report.getMirfReportType()));
        }
        return generatedReport;
    }

    private Table generatePdfForTableReport(DataTableAlgorithmReport report) {

        PdfFont font = null;
        PdfFont bold = null;
        try {
            font = PdfFontFactory.createFont(StandardFonts.HELVETICA);
            bold = PdfFontFactory.createFont(StandardFonts.HELVETICA_BOLD);
        } catch (IOException e) {
            e.printStackTrace();
        }

        var table = new Table(report.getTable().getColumns().size());

        table.setWidth(UnitValue.createPercentValue(100));

        addHeaders(table, report.getTable().getColumns(), bold);
        for (var row : report.getTable().getRows())
            addRow(table, row, font, report.getTable().getColumns());

        return table;
    }

    private void addRow(Table table, Dictionary<String, String> items, PdfFont font, Collection<String> headers) {
        for (var header : headers)
            table.addCell(new Cell().add(new Paragraph(items.get(header)).setFont(font)));
    }

    private void addHeaders(Table table, Collection<String> items, PdfFont font) {
        for (var item : items)
            table.addHeaderCell(new Cell().add(new Paragraph(item).setFont(font)));
    }
}

File: src/main/kotlin/com/mirf/features/reports/pdf/DataTableReportToPdfElementConverter.kt
package com.mirf.features.reports.pdf

import com.itextpdf.io.font.constants.StandardFonts
import com.itextpdf.io.image.ImageDataFactory
import com.itextpdf.io.source.ByteArrayOutputStream
import com.itextpdf.kernel.font.PdfFont
import com.itextpdf.kernel.font.PdfFontFactory
import com.itextpdf.layout.element.Cell
import com.itextpdf.layout.element.Image
import com.itextpdf.layout.element.Paragraph
import com.itextpdf.layout.element.Table
import com.itextpdf.layout.property.UnitValue
import com.mirf.core.data.medimage.ImageSeries
import com.mirf.core.data.report.DataTableAlgorithmReport
import com.mirf.features.reports.PdfElementData
import com.mirf.features.repositoryaccessors.AlgorithmExecutionException
import java.util.*
import javax.imageio.ImageIO


//----------------------------------------------------------------------------------------------------------

fun DataTableAlgorithmReport.asPdfElementData(): PdfElementData {
    val font: PdfFont? = PdfFontFactory.createFont(StandardFonts.HELVETICA)
    val bold: PdfFont? = PdfFontFactory.createFont(StandardFonts.HELVETICA_BOLD)

    val table = Table(this.table.columns.size)

    table.width = UnitValue.createPercentValue(100f)

    addHeaders(table, this.table.columns, bold)
    for (row in this.table.rows)
        addRow(table, row, font, this.table.columns)

    return PdfElementData(table)
}

private fun addRow(table: Table, items: Dictionary<String, String>, font: PdfFont?, headers: Collection<String>) {
    for (header in headers)
        table.addCell(Cell().add(Paragraph(items.get(header)).setFont(font)))
}

private fun addHeaders(table: Table, items: Collection<String>, font: PdfFont?) {
    for (item in items)
        table.addHeaderCell(Cell().add(Paragraph(item).setFont(font)))
}

//----------------------------------------------------------------------------------------------------------

fun ImageSeries.asPdfElementData(range: Iterable<Int> = 0 until this.images.size): PdfElementData {
    val images = this.images.map { x -> x.image }

    val result = Paragraph()
    try {
        for (i in range) {
            val image = images[i]
            val stream = ByteArrayOutputStream()
            ImageIO.write(image, "jpg", stream)

            val pdfImage = Image(ImageDataFactory.create(stream.toByteArray()))
            pdfImage.width = UnitValue.createPercentValue(33f)
            pdfImage.setMargins(5f, 5f, 5f, 5f)
            pdfImage.setHeight(UnitValue.createPercentValue(50f))

            result.add(pdfImage)
        }
    } catch (e: Exception) {
        throw AlgorithmExecutionException(e)
    }

    return PdfElementData(result)
}


